src/algebra/direct_sum_graded.lean

```diff
@@ -543,6 +545,108 @@ end comm_ring
 
 end grade_zero
 
+section to_semiring
+
+variables {R : Type*} [Π i, add_comm_monoid (A i)] [add_monoid ι] [gmonoid A] [semiring R]
+variables {A}
+
+/-- If two ring homomorphisms from `⨁ i, A i` are equal on each `of A i y`,
+then they are equal.
+
+See note [partially-applied ext lemmas]. -/
+@[ext]
+lemma ring_hom_ext' (F G : (⨁ i, A i) →+* R)
+  (h : ∀ i, (F : (⨁ i, A i) →+ R).comp (of _ i) = (G : (⨁ i, A i) →+ R).comp (of _ i)) : F = G :=
+ring_hom.coe_add_monoid_hom_injective $ direct_sum.add_hom_ext' h
+
+section to_move
+
+lemma _root_.add_monoid_hom.ext_iff₂
+  {R S T : Type*} [add_zero_class R] [add_zero_class S] [add_comm_monoid T] {f g : R →+ S →+ T} :
+  f = g ↔ (∀ x y, f x y = g x y) :=
+add_monoid_hom.ext_iff.trans $ forall_congr $ λ _, add_monoid_hom.ext_iff
+
+/-- An `add_monoid_hom` preserves multiplication if pre- and post- composition with
+`add_monoid_hom.mul` are equivalent. By converting the statement into an equality of
+`add_monoid_hom`s, various specialized `ext` lemmas about `→+` can then apply. -/
+lemma _root_.add_monoid_hom.map_mul_iff
+  {R : Type u_1} {S : Type u_2} [semiring R] [semiring S] (f : R →+ S) :
+  (∀ x y, f (x * y) = f x * f y) ↔
+    (f.comp_hom : _ →+ R →+ _).comp add_monoid_hom.mul =
+      ((add_monoid_hom.mul.comp f).flip.comp f).flip :=
+iff.symm add_monoid_hom.ext_iff₂
```

Comment:
I found this was a very useful pattern, but again we don't have any other lemmas of this style. One youtube would be to mirror the baffling array of `linear_map.(lr)comp(lr)` definitions to make this a little easier to state.

If we had a definition to compose two add_monoid_homs as `(my_comp f g) x y = f (g x) (g y)` then most of the  need for this lemma would vanish

Reply:
[can't guess what got autocorrected to youtube]

I regard add_monoid_hom as "category theory lite" -- where we're bundling morphisms so `.comp` is very convenient -- but if push comes to shove we can just check stuff using `ext`. I've seen other times when it's very handy to have abstract `comp`y lemmas so I don't see anything wrong with this, modulo the warning that I've seen `flip` cause universe unification issues in the past because its type has quite an exotic universe unfortunately.

Reply:
> I found this was a very useful pattern, but again we don't have any other lemmas of this style. One ~youtube~ option would be to mirror the baffling array of `linear_map.(lr)comp(lr)` definitions to make this a little easier to state.
> 
> If we had a definition to compose two add_monoid_homs as `(my_comp f g) x y = f (g x) (g y)` then most of the need for this lemma would vanish

I was bitten by this while writing it - `comp_hom` has the same downside, which is why the type annotation and explicit universe quantifiers are needed.

Reply:
I've gone ahead and added the `linear_map`-style lemmas in #7407